### ISSUE TITLE
Elasticsearch 7.0 and interval deprecation, use fixed_interval or calendar_interval instead

### ISSUE THREAD

**Issue by cdalexndr** [grafana/grafana#19911]
**What happened**:
Elasicsearch outputs warnings when the dashboard is refreshed.

**What you expected to happen**:
No warnings.

**How to reproduce it (as minimally and precisely as possible)**:
Dashboard using elasticsearch data provider.

**Anything else we need to know?**:
elastic search log:
`{"type": "deprecation", "timestamp": "2019-10-18T17:57:35,015+0000", "level": "WARN", "component": "o.e.d.s.a.b.h.DateHistogramAggregationBuilder", "cluster.name": "docker-cluster", "node.name": "5b2f2184bb11", "cluster.uuid": "dobimEZ6TW2Max3JFgQHXQ", "node.id": "94L6WwbCQNOB-toDHUrBZg",  "message": "[interval] on [date_histogram] is deprecated, use [fixed_interval] or [calendar_interval] in the future."  }`

**Environment**:
- Grafana version: Grafana v6.3.6 (commit: fdd2117)
- Data source type & version: Elasticsearch 7.3.0
- OS Grafana is installed on: docker image grafana/grafana
- User OS & Browser: Win 10, Firefox 69.0.3 (64-bit)
- Grafana plugins: default
- Others:

**Comment by Theoooooo** [comment]
Looking at the ElasticSearch documentation, it only seems to be a little change in the query being done.
But the first note attached seems to be a bit more complicated than what i though ^^
```
Combined interval field is deprecated
[7.2] Deprecated in 7.2. interval field is deprecatedHistorically both calendar and fixed intervals were configured in a single interval field, which led to confusing semantics. Specifying 1d would be assumed as a calendar-aware time, whereas 2d would be interpreted as fixed time. To get "one day" of fixed time, the user would need to specify the next smaller unit (in this case, 24h).

This combined behavior was often unknown to users, and even when knowledgeable about the behavior it was difficult to use and confusing.

This behavior has been deprecated in favor of two new, explicit fields: calendar_interval and fixed_interval.

By forcing a choice between calendar and intervals up front, the semantics of the interval are clear to the user immediately and there is no ambiguity. The old interval field will be removed in the future. ```

**Comment by Elfo404** [comment]
This is a breaking change which is going to be introduced in 8.0.0 (https://www.elastic.co/guide/en/elasticsearch/reference/master/migrating-8.0.html), we should tackle this ASAP

**Comment by Elfo404** [comment]
in https://github.com/grafana/grafana/pull/41729 we are going to set `fixed_interval` whenever the configured ES version is >= 8.

I would personally do the same for ES 7.2+ as most of the time this was set to `auto` what ends up being used is practically always a valid `fixed_interval` (`ms`, `s`, `m`, `h`, `d`) but this depends on the time range. i assume for a large enough time range this can become `1d` which is treated as a calendar interval (in this case i'd consider this behavior a bug, because in every other "auto" case we always end up with a fixed interval).

the issue is that we allow any value in the field. I'm not exactly sure how many users did that (I believe 0 would be a safe bet) but `w` (week), `M` (month), `q` (quarter) and `y` (year) are all valid value when using `interval` as it gets interpreted as `calendar_interval`. changing it to `calendar interval would break queries for users relying on it.

To summarize, we have 2 options here:

**1. Do Nothing**
We do not fix the deprecation warning. No breaking changes are introduced and the warning will go away when users upgrade to ES 8 as they will use `fixed_interval`. in that case, their queries may break if the specified an interval in the form of the ones specified above (calendar intervals) but upgrading to a new major version of ES this could be expected.

**2. Use fixed_interval starting from 7.2**
like above, but we fix the warning message and we may break queries for users that are not upgrading ES version.

I'd personally go for option 2 but I'd wait for a new major Grafana version (9) to properly document it as a breaking change.

opinions? cc. @grafana/observability-squad

**Comment by aocenas** [comment]
> I would personally do the same for ES 7.2+ as most of the time this was set to auto what ends up being used is practically always a valid fixed_interval (ms, s, m, h, d) but this depends on the time range. i assume for a large enough time range this can become 1d which is treated as a calendar interval (in this case i'd consider this behavior a bug, because in every other "auto" case we always end up with a fixed interval).

To me this actually makes sense. 1 day is a calendar construct so it should be a calendar_interval, if you don't want that you could specify 24h. I think this is sort of convention most of the time when handling intervals.

**Comment by ifrost** [comment]
Could we add a component allowing to select between _fixed_ or _calendar_ interval? `1d` has different meaning when used as _fixed_ (exactly 24 hours) and as _calendar_ interval (starting from midnight). Could we allow the user to choose and default to _fixed_?

I think it makes sense to make changes scheduled for Grafana 9 unless it's becoming a huge pain to users - in that case probably we could make a breaking change in a minor release.

**Comment by Elfo404** [comment]
I think this needs more thought as until now the support for calendar aware intervals was accidental. 
I agree that adding an aggregation level setting would make the most sense, but fully supporting calendar intervals requires a bit more tinkering. Grafana itself only reasons in fixed_intervals (in Grafana `1y` is defined as 365 days (https://github.com/grafana/grafana/blob/baab021fec6b2cab7e54ed574af2a0662752ce5b/pkg/tsdb/intervalv2/intervalv2.go#L16).

My proposal is to start with defaulting to fixed_interval as a baseline (in the next major release) and then introduce support for calendar intervals later on (can also be the same release).

**Comment by aocenas** [comment]
Hmm right I got it now, yeah seems like technically the 2nd proposal makes more sense even though it's less nice. I think it would be better to have both default to fixed and some way to manually switch to calendar at the same release. That way we can at least say this will break for some but there is a way to manually fix it right away.

**Comment by fenchu** [comment]
with ES v8.0 this is no longer just a deprecation: 
`Elasticsearch error: [1:522] [date_histogram] unknown field [interval] did you mean [fixed_interval]?`
I cant change this in grafana, but if I run the query in ES data is fine.

**Comment by gabor** [comment]
with https://github.com/grafana/grafana/pull/50297 merged, we will always use `fixed_interval`. this should better represent Grafana's time-ranges in elasticsearch.
